### PR TITLE
libroach,cli,server: add thread stack dump facility

### PR DESCRIPTION
**Pull request by petermattis**
Add a facility for dumping the stack traces for all threads in the
process under linux. The technique used was adapted from
github.com/thoughtspot/threadstacks. The list of threads is retrieved by
scanning `/proc/self/tasks`. A realtime signal is sent to each thread
using `rt_tgsigqueueinfo`. A custom signal handler for that signal uses
the glibc `backtrace` facility to retrieve the thread's
stack. Communication between the coordinating thread and the signalled
thread is performed using a pipe (most other synchronization primitives
are not safe to use from a signal handler).

Hook up `/debug/threads` endpoint and add a link on the debug page of
the admin UI. Extend `/_status/stacks` to allow the optional retrieval
of thread stacks (vs the default of goroutine stacks). Enhance `debug
zip` to retrieve the thread stacks for each node.

Release note (admin ui change): Improve debuggability of C++-level
issues by providing access to thread stack traces via a new
`/debug/threads` endpoint which is exposed on the Admin UI advanced
debug page. Include thread stack traces in the info collected by `debug
zip`. Thread stack traces are currently only available on Linux.

### CODE REVIEW

**Comment by petermattis** [comment]
This is now ready for a real review.

@knz I'm hoping you can review the `debug zip` changes. If you have knowledge of signals and threads in C++ you can also take a look at `libroach/stack_trace.cc`.

@tbg I'm nominating you to look at `libroach/stack_trace.cc`. If you know someone who would be better, let me know.

@dhartunian Please take a look at the change to `pkg/ui/src/views/reports/containers/debug/index.tsx`. Should be simple enough. I just cargo-culted.

**Comment by petermattis** [comment]
It would be nice to backport this to 19.2, though we have to balance that against the risk that this new endpoint could crash a node, making `debug zip` more dangerous and less useful. Probably should let this bake on master for a little bit.

**Review by dhartunian — APPROVED** [review]
`pkg/ui/src/views/reports/containers/debug/index.tsx` LGTM 👍

File: c-deps/libroach/stack_trace.cc [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  // Signal all threads to write their stack trace in a pre-allocated
+  // area. Note that some threads might have died by now, so
+  // signalling them will fail.
+  std::vector<std::unique_ptr<ThreadStack>> stacks;
+  const auto pid = getpid();
+  const auto uid = getuid();
+  for (auto tid : tids) {
+    const uint64_t blocked = BlockedSignals(tid);
+    if ((blocked & (1ULL << kStackTraceSignal)) != 0) {
+      // The thread is blocking receipt of our signal, so don't bother
```
Comment (tbg) [inline c-deps/libroach/stack_trace.cc]: What does this mean (i.e. when would a thread not accept our signal) and does that mean we're just omitting it? Is that ok?

File: c-deps/libroach/stack_trace.cc [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    auto timeout = end - NowMillis();
+    if (timeout <= 0) {
+      break;
+    }
+    auto ret = poll(pollfds, 1, int(timeout));
+    if (ret == -1) {
+      continue;
+    }
+    if (ret == 0) {
+      // We timed out before reading all of the stacks.
```
Comment (tbg) [inline c-deps/libroach/stack_trace.cc]: Are we returning partial info in this case? Are we marking it as partial?

**Review by tbg — APPROVED** [review]
Definitely 🐶 on the details but the overall approach in `stack_trace.cc` looks good. I can believe how this all works in the happy case, but was curious how we handle threads that don't respond on time or even block the signal. With our luck, these cases will end up mattering...

File: c-deps/libroach/stack_trace.cc (hunk 2 above)
Comment (petermattis) [reply c-deps/libroach/stack_trace.cc]: Done. See the code below where we indicate `(no response)` for a thread which didn't return a stack.

**Review by petermattis — COMMENTED** [review]
TFTR!

@knz I suspect you might have some knowledge of these areas. If you don't, feel free to ignore an I'll rely on @tbg's science dog stamp.

**Review by knz — COMMENTED** [review]
A couple high-level comments:

- I wonder what happens when the internal kernel pipe buffer is full and the program is running with only 1 core, if my reading is right you'd get a deadlock in that case: the sig handler tries to write, but the reader is not scheduled at that time.

- signals are delivered synchronously. My design would be to use a global `volatile` variable to hold the stack. Then iterate the following 3 steps:

   1. send the sig to 1 thread
   2. that thread gets the sig delivered, dumps its backtrace to the global var, sig handler completes
   3. main thread gets control back, picks up the state from the global
   4. start back to 1 for next thread

It's slower but it would avoid the uncertainty of polling on a pipe. The timeout makes me uncomfortable (on a loaded system, where we're more likely to want stacks,, it's also more likely to reach this timeout). 

- I wonder why you use `backtrace` but refrain from using `backtrace_symbols` to get the translation immediately and locally. This would increase the ability of users to self-serve.

- I recall we still support a non-glibc build on linux (musl?). This may need to be tested and the build flags adjusted accordingly.

- I would have the thing return an error if nothing can be read from `/proc/self/tasks`, or if one of the syscalls fails, or when the backtrace call fails,  instead of silently reporting "no threads". Also different errors for every case so we can troubleshoot if it ever happens. This would also clarify for tech support when a particular OS/lib combination should be marked as incompletely supported.

> Reviewed 9 of 14 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-:-M1VEFYv6WX4sg75Ncgu:b-oo7xi2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[c-deps/libroach/stack_trace.cc, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFnPb9vZJbNobBsZw:-M1VFnPb9vZJbNobBsZx:b-egyd6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L67)):*
> ```cpp
> std::vector<pid_t> ListThreads() {
>   std::vector<pid_t> pids;
>   DIR* dir = opendir("/proc/self/task");
> ```

I notice you have an EINTR loop below. Is it worth wrapping the various syscalls, including opendir/readdir/closedir here, with an EINTR retry loop?
___
*[c-deps/libroach/stack_trace.cc, line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFOsX2-T6_gTYPozJ:-M1VFOsX2-T6_gTYPozK:b-1j6tj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L83)):*
> ```cpp
>       continue;
>     }
>     pids.push_back(stoll(child));
> ```

`stoll` would throw an exception if the conversion fails. 
Do we have exception handling for c++ code somewhere? or is it going to crash the process?
Also if an exception is thrown, `closedir` is not called and the process leaks a file descriptor. I would enclose the entire opendir..closedir block in try..catch for this purpose.
___
*[c-deps/libroach/stack_trace.cc, line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VGIXnCeRH5fCk1LLq:-M1VGIXnCeRH5fCk1LLr:b-na3eil) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L91)):*
> ```cpp
> uint64_t BlockedSignals(pid_t tid) {
>   const std::string path = "/proc/" + std::to_string(tid) + "/status";
>   int fd;
> ```

this way to read the file contents does not make me super comfortable: I don't se a strong reason why 1024 bytes are sufficient, and more would blow the stack perhaps?

What about the classic
```
std::string contents;
{
  std::ifstream ifs("/proc/.../status");
  contents = std::string( (std::istreambuf_iterator<char>(ifs) ),
                       (std::istreambuf_iterator<char>()    ) );
}
```
___
*[c-deps/libroach/stack_trace.cc, line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VG_R54Kq6wy8zoJPC:-M1VG_R54Kq6wy8zoJPD:be2mvjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L122)):*
> ```cpp
>   }
>   data = data.substr(pos + needle.size());
>   return stoull(data, nullptr, 16);
> ```

ditto my comment about this throwing execptions
___
*[c-deps/libroach/stack_trace.cc, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VH1uu-PKGJieA2a1x:-M1VH1uv-KRWvxaBzrn4:bkbbkfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L160)):*
> ```cpp
>   // interrupted by the stacktrace collection signal.
>   action.sa_flags = SA_ONSTACK | SA_RESTART | SA_SIGINFO;
>   return sigaction(kStackTraceSignal, &action, nullptr) == 0;
> ```

if memory serves the idiom is a first `sigaction` to retrieve the current flags, then a second `sigaction` to override the particular flags of interest.
Then at the end, to clean up, a last `sigaction` to restore what was there the first time around.

The reason why this matters is that the Go runtime used to (maybe still is) particular about `sa_mask` and, on some platforms (I don't know about linux) `sa_restorer`.
___
*[c-deps/libroach/include/libroach.h, line 620 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VE4lNCHABVPFEZjen:-M1VE4lNCHABVPFEZjeo:b-h8rodz) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/include/libroach.h#L620)):*
> ```objc
> // atos on Darwin) to symbolize.
> DBString DBDumpThreadStacks();
>   
> ```

nit: two stray space characters

**Comment by knz** [comment]
Sorry no the signal is not synchronous because it's a different thread. Obviously I was mistaken.

Still I don't think the pipe is mandatory, a global volatile var and a spinlock would also probably do.

**Review by petermattis — COMMENTED** [review]
TFTR, @knz!

> I wonder what happens when the internal kernel pipe buffer is full and the program is running with only 1 core, if my reading is right you'd get a deadlock in that case: the sig handler tries to write, but the reader is not scheduled at that time.

Since linux 2.6.11, the pipe buffer defaults to 64KB. Prior to that, the pipe buffer was 4KB. I doubt we'd ever experience the pipe buffer becoming full. But even if that happens, I'm not convinced you'd get a deadlock regardless. Your second comment about the signals being delivered asynchronously might be recognition of this. Even though there is only 1 core, if you have multiple threads, when one thread writes to the pipe, another one would be reading. 

> signals are delivered synchronously. My design would be to use a global volatile variable to hold the stack. Then iterate the following 3 steps:
> 
> 1. send the sig to 1 thread
> 2. that thread gets the sig delivered, dumps its backtrace to the global var, sig handler completes
> 3. main thread gets control back, picks up the state from the global
> 4. start back to 1 for next thread

The global variable approach avoids heap allocation of the `ThreadStack` structure, but what do we do if one of the threads doesn't respond (e.g. because the signal is blocked). With the heap allocation approach you can just leak the structure. With a global variable you can't.

> It's slower but it would avoid the uncertainty of polling on a pipe. The timeout makes me uncomfortable (on a loaded system, where we're more likely to want stacks,, it's also more likely to reach this timeout).

FWIW, grabbing the stacks for 30 threads takes 2-10ms on my Mac laptop when running under Docker. We could increase the length of the timeout, but because we can't guarantee signal delivery we'll always need some timeout. Do you have a better suggestion than 5s?

> I wonder why you use backtrace but refrain from using backtrace_symbols to get the translation immediately and locally. This would increase the ability of users to self-serve.

No strong reason. I saw some scary stuff about `backtrace_symbols` somewhere on the nets. Do you have experience with it? Also, it doesn't provide line number information was is irritating, but I suppose symbols are better than no-symbols. Done.

> I recall we still support a non-glibc build on linux (musl?). This may need to be tested and the build flags adjusted accordingly.

Good catch. Do you recall why we support musl? I can't recall anyone ever using it, and supporting it without cause is irritating. I've added some additional `#ifdefs` and verified that `build/builder.sh mkrelease linux-musl` works.

> I would have the thing return an error if nothing can be read from /proc/self/tasks, or if one of the syscalls fails, or when the backtrace call fails, instead of silently reporting "no threads". Also different errors for every case so we can troubleshoot if it ever happens. This would also clarify for tech support when a particular OS/lib combination should be marked as incompletely supported.

It looks like `backtrace` can't fail. The [man page](http://man7.org/linux/man-pages/man3/backtrace.3.html) doesn't say anything about a failure return value. I've added some additional error handling, such as `ListThreads` return nothing. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-:-M1krpbM7i1pCKG9Ab_l:b-mjua44)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[c-deps/libroach/stack_trace.cc, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFnPb9vZJbNobBsZw:-M1l4cUj9jGkjai5so1q:bngt21m) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L67)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I notice you have an EINTR loop below. Is it worth wrapping the various syscalls, including opendir/readdir/closedir here, with an EINTR retry loop?
</blockquote></details>

`opendir/readdir/closedir` are C library functions, not system calls. My internet searches don't reveal them being wrapped in EINTR retry loops. Do you know differently?
___
*[c-deps/libroach/stack_trace.cc, line 83 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFOsX2-T6_gTYPozJ:-M1l59999ProSBXX8_O1:b-5oy6n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L83)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`stoll` would throw an exception if the conversion fails. 
Do we have exception handling for c++ code somewhere? or is it going to crash the process?
Also if an exception is thrown, `closedir` is not called and the process leaks a file descriptor. I would enclose the entire opendir..closedir block in try..catch for this purpose.
</blockquote></details>

We haven't used C++ exceptions anywhere and they aren't used in RocksDB. I'd prefer not to introduce them here. I've switched to using `strtoll` and elsewhere `strtoull`.
___
*[c-deps/libroach/stack_trace.cc, line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VGIXnCeRH5fCk1LLq:-M1l7EU_Dwp_vSFuaSwj:bpjh1x6) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L91)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this way to read the file contents does not make me super comfortable: I don't se a strong reason why 1024 bytes are sufficient, and more would blow the stack perhaps?

What about the classic
```
std::string contents;
{
  std::ifstream ifs("/proc/.../status");
  contents = std::string( (std::istreambuf_iterator<char>(ifs) ),
                       (std::istreambuf_iterator<char>()    ) );
}
```

</blockquote></details>

The loop reads the file no matter how large it is, it only does so in 1024 byte increments. The file is generally smaller than that. I've never been comfortable with the C++ IO idioms. 
___
*[c-deps/libroach/stack_trace.cc, line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VG_R54Kq6wy8zoJPC:-M1l7OviCzEFh9PTCguS:b-e9q4au) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L122)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto my comment about this throwing execptions
</blockquote></details>

Switched to `strtoull`.
___
*[c-deps/libroach/stack_trace.cc, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VH1uu-PKGJieA2a1x:-M1l7V7IA0QZ-ElxX-YE:b-q9xl30) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L160)):*
> if memory serves the idiom is a first sigaction to retrieve the current flags, then a second sigaction to override the particular flags of interest.
Then at the end, to clean up, a last sigaction to restore what was there the first time around.

I've restructured this code so we restore the old signal handler on return.

> The reason why this matters is that the Go runtime used to (maybe still is) particular about sa_mask and, on some platforms (I don't know about linux) sa_restorer.

Can you point me towards where this is (or was) being done in the Go runtime? When I looked around I didn't see anything that was problematic, but the signal handling is complex and I might have missed something. Note that `sa_mask` is empty here.

**Review by knz — COMMENTED** [review]
This looks about good to go. 

> Do you have a better suggestion than 5s?

No better suggestion, no. We can also just try your current approach out and see where that brings us during the release cycle.

> backtrace_symbols ... Do you have experience with it? 

Casual experience. It worked in a small experiment I once wrote. I don't know about real-world. I can imagine it could break if we ran `strip` on executables.

>  Do you recall why we support musl? 

Yes, this is the only way we currently have to produce statically linked binaries on linux. We had early customers that wanted/asked for this.

I also have a few leftover comments see below.

> Reviewed 10 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-:-M1ydC4Q3m8nrXlxu3PF:bm4ufg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @tbg)

___
*[c-deps/libroach/db.cc, line 813 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1yZiUH5ayFLLqmrwHy:-M1yZiUI18YE9FxXvZOD:b-48h2yq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5d680225d351c37324519249cf0a12b8d6c5a7/c-deps/libroach/db.cc#L813)):*
> ```cpp
> DBStatus DBGetStats(DBEngine* db, DBStatsResult* stats) {
>   auto stacks = DumpThreadStacks();
>   fprintf(stderr, "%s\n", stacks.c_str());
> ```

Did you intend to keep this?
___
*[c-deps/libroach/stack_trace.cc, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFnPb9vZJbNobBsZw:-M1yZrFp6BMJhsDKZe46:b25ldb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L67)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`opendir/readdir/closedir` are C library functions, not system calls. My internet searches don't reveal them being wrapped in EINTR retry loops. Do you know differently?
</blockquote></details>

I went and check at the source code of GNU libc and BSD libc and neither contains EINTR retry loops on the directory functions. They _do_ use open/read/etc under the hood though. Obviously given these are not real files, the time window at which a signal could be received is very small, but the race condition is still theoretically there.
___
*[c-deps/libroach/stack_trace.cc, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VH1uu-PKGJieA2a1x:-M1ybJEk9Xqcngw-TykX:b-u5wyel) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L160)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> if memory serves the idiom is a first sigaction to retrieve the current flags, then a second sigaction to override the particular flags of interest.
Then at the end, to clean up, a last sigaction to restore what was there the first time around.

I've restructured this code so we restore the old signal handler on return.

> The reason why this matters is that the Go runtime used to (maybe still is) particular about sa_mask and, on some platforms (I don't know about linux) sa_restorer.

Can you point me towards where this is (or was) being done in the Go runtime? When I looked around I didn't see anything that was problematic, but the signal handling is complex and I might have missed something. Note that `sa_mask` is empty here.
</blockquote></details>

Here's the code in os_linux.go:

```go
//go:nosplit
//go:nowritebarrierrec
func setsig(i uint32, fn uintptr) {
        var sa sigactiont
        sa.sa_flags = _SA_SIGINFO | _SA_ONSTACK | _SA_RESTORER | _SA_RESTART
        sigfillset(&sa.sa_mask)
        // Although Linux manpage says "sa_restorer element is obsolete and
        // should not be used". x86_64 kernel requires it. Only use it on
        // x86.
        if GOARCH == "386" || GOARCH == "amd64" {
                sa.sa_restorer = funcPC(sigreturn)
        }
...
```

So as you see 1) there's sa_restorer funkiness 2) sa_flags is different from yours.

Here's another example thing that was happening in `sys_darwin.go`:

```go
//go:nosplit
//go:nowritebarrierrec
func setsig(i uint32, fn uintptr) {
        var sa sigactiont
        sa.sa_flags = _SA_SIGINFO | _SA_ONSTACK | _SA_RESTART
        sa.sa_mask = ^uint32(0)
        sa.sa_tramp = unsafe.Pointer(funcPC(sigtramp)) // runtime·sigtramp's job is to call into real handler
        *(*uintptr)(unsafe.Pointer(&sa.__sigaction_u)) = fn
        sigaction(i, &sa, nil)
}
```

This approach was actually catastrophic, because on Darwin it's not actually possible to _retrieve_ the current value of `sa_tramp` with `sigaction()`, so the general idea of "get current settings, customize, then restore" does not work. This was biting us very hard in go-libedit.

I think this was changed in go 1.12/1.13 though for Darwin. 
___
*[c-deps/libroach/stack_trace.cc, line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1yalJW6RXABhm1swRs:-M1yalJW6RXABhm1swRt:b34wutu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5d680225d351c37324519249cf0a12b8d6c5a7/c-deps/libroach/stack_trace.cc#L127)):*
> ```cpp
>   const std::string needle("SigBlk:");
>   size_t pos = data.find(needle);
>   if (pos == data.npos) {
> ```

I would certainly recommend reporting an error in this case. This sounds like a serious condition which we'd need to hear about (even perhaps record it in telemetry in Go).

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[c-deps/libroach/db.cc, line 813 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1yZiUH5ayFLLqmrwHy:-M1zL8Im0rGQOVUhOOjC:blwok1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5d680225d351c37324519249cf0a12b8d6c5a7/c-deps/libroach/db.cc#L813)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Did you intend to keep this?
</blockquote></details>

Oops. No. Removed.
___
*[c-deps/libroach/stack_trace.cc, line 67 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VFnPb9vZJbNobBsZw:-M1zLFueCqARmWnyU7_f:b-q7hvug) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L67)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I went and check at the source code of GNU libc and BSD libc and neither contains EINTR retry loops on the directory functions. They _do_ use open/read/etc under the hood though. Obviously given these are not real files, the time window at which a signal could be received is very small, but the race condition is still theoretically there.
</blockquote></details>

Agreed the race condition is there. I've wrapped these calls in EINTR retry loops, and a few other places where system calls were being done without retry loops.
___
*[c-deps/libroach/stack_trace.cc, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1VH1uu-PKGJieA2a1x:-M1zPn9P0Hd5yz4Rgd5h:btdkz2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/286e5841b8c1a5771fc408f281a2604efed47138/c-deps/libroach/stack_trace.cc#L160)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here's the code in os_linux.go:

```go
//go:nosplit
//go:nowritebarrierrec
func setsig(i uint32, fn uintptr) {
        var sa sigactiont
        sa.sa_flags = _SA_SIGINFO | _SA_ONSTACK | _SA_RESTORER | _SA_RESTART
        sigfillset(&sa.sa_mask)
        // Although Linux manpage says "sa_restorer element is obsolete and
        // should not be used". x86_64 kernel requires it. Only use it on
        // x86.
        if GOARCH == "386" || GOARCH == "amd64" {
                sa.sa_restorer = funcPC(sigreturn)
        }
...
```

So as you see 1) there's sa_restorer funkiness 2) sa_flags is different from yours.

Here's another example thing that was happening in `sys_darwin.go`:

```go
//go:nosplit
//go:nowritebarrierrec
func setsig(i uint32, fn uintptr) {
        var sa sigactiont
        sa.sa_flags = _SA_SIGINFO | _SA_ONSTACK | _SA_RESTART
        sa.sa_mask = ^uint32(0)
        sa.sa_tramp = unsafe.Pointer(funcPC(sigtramp)) // runtime·sigtramp's job is to call into real handler
        *(*uintptr)(unsafe.Pointer(&sa.__sigaction_u)) = fn
        sigaction(i, &sa, nil)
}
```

This approach was actually catastrophic, because on Darwin it's not actually possible to _retrieve_ the current value of `sa_tramp` with `sigaction()`, so the general idea of "get current settings, customize, then restore" does not work. This was biting us very hard in go-libedit.

I think this was changed in go 1.12/1.13 though for Darwin. 
</blockquote></details>

Thanks for the pointers to code. The only difference in the flags is the presence of `SA_RESTORER`. As far as I can tell, `SA_RESTORER` is something internal to glibc and will be populated by glibc. See http://man7.org/linux/man-pages/man2/sigreturn.2.html. The Go runtime is doing this itself because it is making direct system calls. 
___
*[c-deps/libroach/stack_trace.cc, line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-M1yalJW6RXABhm1swRs:-M1zOdeI5AjCHuk_RImv:b-i81vi1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a5d680225d351c37324519249cf0a12b8d6c5a7/c-deps/libroach/stack_trace.cc#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would certainly recommend reporting an error in this case. This sounds like a serious condition which we'd need to hear about (even perhaps record it in telemetry in Go).
</blockquote></details>

Agreed this would be very surprising. I've added error messaging to this function. Actually recording this in telemetry would be extremely onerous, but at least we'll see the error in the stack dump.

**Review by knz — APPROVED** [review]
:lgtm:

> Reviewed 1 of 11 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45321#-:-M1zZ3f8-y44HxSunTGv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

**Comment by petermattis** [comment]
TFTR, @knz and @tbg! If we see any problems with this during the stability period, I'm more than willing to revert.

**Pull request merged**